Add documentation for custom group roles (#8375)
PrairieLearn/PrairieLearn: docs/assessment.md

Students are able to see their groupmates' UIDs, which can become a point of contact to communicate with eachother outside of PrairieLearn. They are also able to leave their group to join a different one.

## Enabling custom group roles for collaborative assessments

By default, students working in a collaborative group assessments can view and submit every question. However, instructors can define **custom group roles**, which can be assigned different permissions to facilitate role-based teamwork. Roles can receive the following permissions:

- Submitting specific questions
- Viewing specific questions

> - Assigning group roles before & during an assessment

Reviewer: ```suggestion
- Assigning group roles for other students
```

> Students are allowed to take on multiple roles. This may be necessary when users leave and join groups after assessments have started.

Reviewer: ```suggestion
Although in most cases each student is expected to take one role, students are allowed to take on multiple roles in some narrow scenarios, such as when users leave and join groups after assessments have started.
```

To opt-in to custom group roles, group roles must be defined at the root of the `infoAssessment.json` file. For example:

```json
{
  "groupRoles": [
    {
      "name": "Manager",
      "minimum": 1,
      "maximum": 1,

>       "canAssignRolesAtStart": true

Reviewer: ```suggestion
      "canAssignRoles": true
```

Assuming #9290 is merged.

    },
    {
      "name": "Recorder",
      "minimum": 1,
      "maximum": 1
    },
    {
      "name": "Reflector",
      "minimum": 1,
      "maximum": 1
    },
    {
      "name": "Contributor"
    }
  ]
}
```

| Attribute               | Type    | Default | Description                                                              |
| ----------------------- | ------- | ------- | ------------------------------------------------------------------------ |
| `name`                  | string  | -       | The name of the role.                                                    |

> | `minimum`               | integer | 0       | The minimum required number of this role in the assessment.              |
> | `maximum`               | integer | -       | The maximum required number of this role in the assessment.              |

Reviewer: ```suggestion
| `minimum`               | integer | 0       | The minimum required number of students holding this role in the assessment.              |
| `maximum`               | integer | -       | The maximum required number of students holding this role in the assessment.              |
```

> | `canAssignRolesAtStart` | boolean | false   | Allow students with this role to assign roles throughout the assessment. |

Reviewer: ```suggestion
| `canAssignRoles` | boolean | false   | Allow students with this role to assign roles to other students. |
```

Will require a re-formatting of the table.

> Students select their roles before starting an assessment, but they can change their roles mid-assessment if needed. As a safeguard against invalid role configurations, PrairieLearn prevents all question submissions if a group's role configuration does not meet the instructor's specification.

Reviewer: ```suggestion
Students typically select their roles before starting an assessment, but they can change their roles mid-assessment if needed. As a safeguard against invalid role configurations, PrairieLearn prevents students from viewing questions if a group's role configuration does not meet the instructor's specification.
```

### Adding permissions for an assessment

Permissions can be configured at the _assessment_, _zone_, or _question_ level.

The schema for permissions is defined as follows:

```json
{
  "canView": ["Manager", "Reflector", "Recorder", "Contributor"],
  "canSubmit": ["Recorder"]
}
```

| Attribute   | Type            | Default | Description                                                     |
| ----------- | --------------- | ------- | --------------------------------------------------------------- |
| `canView`   | Array of string | -       | The names of roles that can view this part of the assessment.   |
| `canSubmit` | Array of string | -       | The names of roles that can submit this part of the assessment. |

Setting either attribute to `[]` (empty array) means that **no role** can view/submit that part. Setting either attribute to `null` means that **every role** can view/submit that part.

> Permissions defined at a higher level are propagated down the assessment hierarchy (assessment -> zone -> question), but permissions defined at lower levels will override those from the higher level. For example, if the permissions in the example above are configured for the assessment, but a question's permissions are configured to `canSubmit: ["Reflector"]`, then only the _Reflector_ can submit that particular question.

Reviewer 2: It might be nice to show a stripped-down example of an assessment config with a concrete demonstration of setting permissions at the various levels.

Reviewer: Done in ef84fc5dab11a3a9468e32aa7aebdf22ca2664c6.

> Permissions defined at a higher level are propagated down the assessment hierarchy (assessment -> zone -> question), but permissions defined at lower levels will override those from the higher level. For example, if the permissions in the example above are configured for the assessment, but a question's permissions are configured to `canSubmit: ["Recorder"]`, then only the _Recorder_ can submit that particular question.

Reviewer 3: the example above already has only "Recorder" with the permission to Submit. either change the example, or make this question level permission different

### Students joining assessments with custom group roles

> When students join a group, they are automatically assigned a role. Students can always view the roles of other students in the group, both before and during an assessment. More information about the assessment's group roles can be expanded in the dropdown menu.

Reviewer 2: This is not a dropdown menu in the usual sense - maybe say something like this instead?

```suggestion
When students join a group, they are automatically assigned a role. Students can always view the roles of other students in the group, both before and during an assessment. Students can click "View role info" to see more information about the assessment's group roles.
```

You'll need to change the usage of "dropdown menu" a few lines down as well.

![Joining a group assessment with custom group roles](grouproles_join.png)

When expanded, the dropdown menu shows information about the roles, such as the min/max number of assignments and whether a role can assign other roles.

![Group info from student perspective](grouproles_groupinfo.png)

Any student with an assigner role can view additional controls to change the roles of other users in the group. Students with assigner roles can re-assign group roles both before and during an assessment.

![Group role assignment controls](grouproles_assign_roles.png)

### Permission enforcement during assessment

When an instructor restricts the viewing of a question to certain roles, users without those roles will be unable to view that question.

> ![Question cannot be viewed due to group role](grouproles_view_question.png)

Reviewer 3: can this image be smaller?

Author: I'm not sure if you'd like the literal image size to be smaller, or if the screenshot should be less zoomed in. I did the former, but let me know if you meant the latter or something else entirely!

Reviewer 3: I would prefer less zoom (the size of the fonts in the screenshot should be similar to the ones in the text). As it is now, the font size of the image is substantially larger than everything else.

Reviewer: @mfsilva22 can you check https://prairielearn--8375.org.readthedocs.build/en/8375/assessment/#assigning-roles-to-students to see if you're ok with the Zoom level used in these new screenshots?

Additionally, when an instructor restricts the submitting of a question to certain roles, users without those roles will be unable to submit that question.

> ![Question cannot be submitted due to group role](grouproles_submit_question.png)

Reviewer 3: this image still has the "Save" button enabled. Is this what was decided at the end? I thought Geoffrey Herman suggested that both buttons should be disabled.

Author: No, you're correct in that we decided in the end that both buttons should be disabled. I used an outdated screenshot; I'll update that now. Thanks for pointing it out!

### Invalid role configurations

> When a role configuration is invalid, which may occur when a user leaves the group or a new user joins the group, **question submissions are prevented** and users receive visual feedback on how to fix the role configuration.

Reviewer: ```suggestion
When a role configuration is invalid, which may occur when a user leaves the group or a new user joins the group, students become unable to see any questions until the roles are reviewed.
```

| `canAssignRoles` | boolean | false   | Allow students with this role to assign roles to other students.             |

Students typically select their roles before starting an assessment, but they can change their roles mid-assessment if needed. As a safeguard against invalid role configurations, PrairieLearn prevents students from viewing questions if a group's role configuration does not meet the instructor's specification.

#### Adding permissions for an assessment

Permissions can be configured at the _assessment_, _zone_, or _question_ level.

The schema for permissions is defined as follows:

```json
{
  "canView": ["Manager", "Reflector", "Recorder", "Contributor"],
  "canSubmit": ["Recorder"]
}
```

| Attribute   | Type            | Default | Description                                                     |
| ----------- | --------------- | ------- | --------------------------------------------------------------- |
| `canView`   | Array of string | -       | The names of roles that can view this part of the assessment.   |
| `canSubmit` | Array of string | -       | The names of roles that can submit this part of the assessment. |

> Setting either attribute to `[]` (empty array) means that **no role** can view/submit that part. Setting either attribute to `null` means that **every role** can view/submit that part.

Reviewer 2: I think this is true, and it's probably helpful to include so people don't think they need to set both these properties everywhere.

```suggestion
Setting either attribute to `[]` (empty array) means that **no role** can view/submit that part. Setting either attribute to `null` or omitting it entirely means that **every role** can view/submit that part.
```

Reviewer: I fear that this may cause confusion if this is considered in the context of a question where a default is set for a zone/assessment. I'll rewrite this to make it clearer.

Reviewer 2: I realized shortly after submitting this review that yeah, this behavior depends on what's configured at parent levels. Given that, I think this is fine as-is, though I'm happy to review a revised version if you think you can make it more clear.

Reviewer: Rephrased in 956f8c47549aa9d8b45ff351078fc0f49da3fdb3.

| `canAssignRoles` | boolean | false   | Allow students with this role to assign roles to other students.             |

Students typically select their roles before starting an assessment, but they can change their roles mid-assessment if needed. As a safeguard against invalid role configurations, PrairieLearn prevents students from viewing questions if a group's role configuration does not meet the instructor's specification.

#### Adding permissions for an assessment

Permissions can be configured at the _assessment_, _zone_, or _question_ level.

The schema for permissions is defined as follows:

```json
{
  "canView": ["Manager", "Reflector", "Recorder", "Contributor"],
  "canSubmit": ["Recorder"]
}
```

| Attribute   | Type            | Default | Description                                                     |
| ----------- | --------------- | ------- | --------------------------------------------------------------- |
| `canView`   | Array of string | -       | The names of roles that can view this part of the assessment.   |
| `canSubmit` | Array of string | -       | The names of roles that can submit this part of the assessment. |

> Setting either attribute to `[]` (empty array) means that **no role** can view/submit that part. If a question does not set the value of either attribute, it means that **every role** can view/submit that part.

Reviewer 2: Nit: removing the word "question" since these attributes can be set (or not) at any level.

```suggestion
Setting either attribute to `[]` (empty array) means that **no role** can view/submit that part. If neither attribute is set, it means that **every role** can view/submit that part.
```

Reviewer 2: Maybe this is worse, but my CS brain interpreted the original the wrong way (though it's obviously the wrong interpretation if you think about it for 2 seconds):

```suggestion
Setting either `canView` or `canSubmit` to `[]` (empty array) means that **no role** can view or submit that part, respectively. If a question does not set the value of either attribute, it means that **every role** can view/submit that part.
```

Reviewer 2: (I'll leave it up to you to merge my two suggestions 🙂)

Reviewer: ```suggestion
Setting either `canView` or `canSubmit` to `[]` (empty array) means that **no role** can view or submit that part of the assessment, respectively. If either attribute is not set, it means that **every role** has the permission associated to the attribute, i.e., any student with any role can view or submit that part of the assessment.
```
